python/dgl/backend/pytorch/sparse.py

```diff
@@ -691,6 +693,62 @@ def backward(ctx, dB_weights):
         return None, csrmask(gidxB, dB_weights, gidxA), None
 
 
+class SEGMENTMM(th.autograd.Function):
+    @staticmethod
+    @custom_fwd(cast_inputs=th.float16)
+    def forward(ctx, A, B, seglen_A):
+        if A.shape[0] != th.sum(seglen_A):
+            raise Exception("The summation of the elements of seglen_A must be equal to " +
+                "dimension 0 of A. Expected "+ str(A.shape[0]) + "got" + str(th.sum(seglen_A)))
+        if B.dim() != 3:
+            raise Exception("Expected dimension of B is 3. Got " + str(B.dim()))
+        # Reshaping B form 3D to 2D
+        B_3D_shape = B.shape
+        B = B.reshape(B.shape[0] * B.shape[1], B.shape[2])
+        C = th.zeros((A.shape[0], B.shape[1]), device=A.device, dtype=A.dtype)
+        C = _segment_mm(A, B, C, seglen_A)
+        ctx.backward_cache = A, B, seglen_A, B_3D_shape
+        return C
+
+    @staticmethod
+    def backward(ctx, dZ):
+        A, B, seglen_A, B_3D_shape = ctx.backward_cache
+        #  Compute A_grad = Out_grad * B^T
+        A_grad = th.zeros(A.shape, device=A.device, dtype=A.dtype)
+        A_grad = _segment_mm(dZ, B, A_grad, seglen_A, b_trans=True)
+        #  Compute B_grad = A^T * Out_grad
+        B_grad = th.zeros(B.shape, device=B.device, dtype=B.dtype)
+        B_grad = _segment_mm(A, dZ, B_grad, seglen_A, a_trans=True)
+        B_grad = B_grad.reshape(B_3D_shape[0], B_3D_shape[1], B_3D_shape[2])
+        return A_grad, B_grad, None, None, None, None, None, None
+
+
+class GATHERMM(th.autograd.Function):
+    @staticmethod
+    @custom_fwd(cast_inputs=th.float16)
+    def forward(ctx, A, B, idx_a, idx_b):
+        if B.dim() != 3:
+            raise Exception("Expected dimension of B is 3. Got " + str(B.dim()))
+        # Reshaping B form 3D to 2D
+        B_3D_shape = B.shape
+        B = B.reshape(B.shape[0] * B.shape[1], B.shape[2])
+        C = th.zeros((A.shape[0], B.shape[1]), device=A.device, dtype=A.dtype)
+        C = _gather_mm(A, B, C, B_3D_shape[0], idx_a, idx_b)
+        ctx.backward_cache = A, B, idx_a, idx_b, B_3D_shape
+        return C
+
+    @staticmethod
+    def backward(ctx, dZ):
+        A, B, idx_a, idx_b, B_3D_shape = ctx.backward_cache
+        #  Compute A_grad = Out_grad * B^T
+        A_grad = th.zeros(A.shape, device=A.device, dtype=A.dtype)
+        A_grad = _gather_mm_scatter(dZ, B, A_grad, B_3D_shape[0], idx_b=idx_b, b_trans=True)
+        #  Compute B_grad = A^T * Out_grad
+        B_grad = th.zeros(B.shape, device=B.device, dtype=B.dtype)
+        B_grad = _gather_mm_scatter(A, dZ, B_grad, B_3D_shape[0], idx_c=idx_b, a_trans=True)
```

Comment:
Should it be:
`B_grad = _gather_mm_scatter(A, dZ, B_grad, B_3D_shape[0], idx_a=idx_a, idx_c=idx_b, a_trans=True)`?